File: silt-ts/src/resource.ts
import { Signal } from './primitives';

export class Resource<T, Args extends any[] = []> {
    loading = new Signal<boolean>(false);
    error = new Signal<Error | null>(null);
    data: Signal<T | undefined>;
    
    private fetcher: (...args: Args) => Promise<T>;

    constructor(fetcher: (...args: Args) => Promise<T>, init?: T) {
        this.fetcher = fetcher;
        this.data = new Signal<T | undefined>(init);
    }

    async fetch(...args: Args): Promise<T | void> {
        this.loading.set(true);
        this.error.set(null);
        try {
            const res = await this.fetcher(...args);
            this.data.set(res);
            return res;
        } catch (e: any) {
            this.error.set(e);
        } finally {
            this.loading.set(false);
        }
    }
    
    get(): T | undefined {
        return this.data.get();
    }
}

File: silt-python/silt/resource.py
import asyncio
from .primitives import Signal, Node
from .engine import link, eng, propagate

class Resource(Node):
    def __init__(self, fetcher):
        super().__init__()
        self.fetcher = fetcher
        self.data = Signal(None)
        self.loading = Signal(False)
        self.error = Signal(None)

    def fetch(self, *args):
        self.loading.set(True)
        self.error.set(None)
        
        coro = self.fetcher(*args)
        
        async def wrap():
            try:
                res = await coro
                self.data.set(res)
            except Exception as e:
                self.error.set(e)
            finally:
                self.loading.set(False)
                
        asyncio.create_task(wrap())

    def get(self):
        link(self)
        return self.data.get()